source-academy/frontend
# Add file system view desktop user interface

### Description

Add the file system view for the desktop user interface. Similar to most code editors/IDEs available today, the file system view can be found within a sidebar on the left edge of the screen. This sidebar is built to be generic so that future features such as git integration can also make use of it.

*Note: the sidebar tab is now named `Files` instead of `Project` after being deemed to be confusing but the below screenshots are not updated since everything else is still the same.*

![image](https://user-images.githubusercontent.com/5585517/194744795-e010a77f-760e-4dad-bd26-0b7904305e23.png)
![vmconnect_9M3ear1SQ1](https://user-images.githubusercontent.com/5585517/194745228-ffbee7ff-8ec6-4bb8-b47e-e3d29f029605.gif)

Some of the move complicated design decisions were documented in the wiki, namely the [handling of file system updates](https://github.com/source-academy/frontend/wiki/File-System#handling-file-system-updates) and the [lack of options in the BrowserFS API](https://github.com/source-academy/frontend/wiki/File-System#lack-of-options).

This PR adds a new dependency on [react-menu](https://github.com/szhsin/react-menu) due to it being a highly customisable React library for handling context menus (i.e., right click menus) which are used a lot in the file system view. @martin-henz has previously approved the use of this dependency.

Part of #2176.

### Type of change


### How to test

Open the playground and play around with the user interface. Note that the editor is not linked to the file system yet and will not change its contents based on interactions with the file system view.

### Changes that will be tackled in a future PR

* Mobile user interface.
* Resizing of sidebar.
  * Blocked by #2234.
* Different icons based on file extension.
  * Very low priority.
* Hooking up the file system view with the code editor.
* Removing the flickering whenever a file system operation is carried out. This is because the BrowserFS operations are asynchronous.
  * Low priority.
